particle-iot/device-os
# Secures DCT initialization from getting interrupted between creating DCT file and filling it with 0xff to default state

### Description

As title says. Also refactors things a bit to have a common `dct_file.h` implementation shared between Gen 3 / P2 / Tracker M.

### Steps to Test

Force the reset:

```c++
diff --git a/services/inc/dct_file.h b/services/inc/dct_file.h
index 7bd291911..e6f905dc6 100644
--- a/services/inc/dct_file.h
+++ b/services/inc/dct_file.h
@@ -25,6 +25,8 @@
 #include <algorithm>
 #include <mutex>
 
+#include "core_hal.h"
+
 namespace particle {
 namespace dct {
 
@@ -133,19 +135,25 @@ private:
         int r = lfs_mkdir(lfs(), "/sys");
         SPARK_ASSERT((r == 0 || r == LFS_ERR_EXIST));
 
+        lfs_remove(lfs(), "/sys/dct.bin");
+
         /* Check that /sys/dct.bin exists */
         struct lfs_info info;
         r = lfs_stat(lfs(), path_, &info);
 
         int flags = LFS_O_RDWR;
 
-        if (r || info.size == 0) {
+        if (r /*|| info.size == 0*/) {
             /* Create as well if doesn't exist or was just created and not written */
             flags |= LFS_O_CREAT;
         }
 
         SPARK_ASSERT(open(flags));
 
+        if (flags & LFS_O_CREAT) {
+            HAL_Core_Enter_Bootloader(0);
+        }
+
         if (flags & LFS_O_CREAT) {
             LOG_DEBUG(INFO, "Initializing empty DCT");
             /* Fill with 0xff for compatibility with raw flash DCD */
```

The device will reset into DFU without blinking LED. Flash vanilla 5.0.0 or `develop`. It will boot-up with no LEDs etc and broken DCT filled with zeros.

### Example App

N/A

### References

N/A

---

### Completeness
